### PR TITLE
PERL-791 Change stream support

### PR DESCRIPTION
**Pull request by phaylon**
Implements PERL-791 "Change stream support"

Overview:
* Introduces `MongoDB::ChangeStream` as the reconnecting wrapper around change stream aggregations returned from `$collection->watch`.
* Adjusts `::_CommandCursorOp` for tailable `::Op::_Aggregate` instances.
* Adjusts `::_Aggregate` to deal with `maxAwaitTimeMS`.
* Adds `watch` to `::Collection`.
* Adds `MongoDB::InvalidOperationError`.
* Tests for default usage, `maxAwaitTimeMS`, `resumeAfter`, `fullDocument`.

### CODE REVIEW

File: lib/MongoDB/ChangeStream.pm [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    isa => Str,
+    predicate => '_has_full_document',
+);
+
+has _resume_token => (
+    is => 'rw',
+    init_arg => 'resume_after',
+    predicate => '_has_resume_token',
+    lazy => 1,
+    builder => '_build_resume_token',
```
Comment (xdg) [inline lib/MongoDB/ChangeStream.pm]: Is there a builder?

File: lib/MongoDB/ChangeStream.pm [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    clearer => '_clear_cursor',
+);
+
+has collection => (
+    is => 'ro',
+    isa => MongoDBCollection,
+    required => 1,
+);
+
+has options => (
```
Comment (xdg) [inline lib/MongoDB/ChangeStream.pm]: Please rename "aggregation_options" to make it clear these are not the exact same as the options passed to `watch`.

File: lib/MongoDB/ChangeStream.pm [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    my ($self) = @_;
+
+    # starting point is construction time instead of first next call
+    $self->_cursor;
+}
+
+sub _build_cursor {
+    my ($self) = @_;
+
+    my $pipeline = $self->pipeline;
```
Comment (xdg) [inline lib/MongoDB/ChangeStream.pm]: Is this extraneous?

File: lib/MongoDB/ChangeStream.pm [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+    my @pipeline = @{ $self->pipeline || [] };
```
Comment (xdg) [inline lib/MongoDB/ChangeStream.pm]: I think the `|| []` isn't necessary anymore given `required` and the type constraint.

File: lib/MongoDB/ChangeStream.pm [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+            ),
+        }},
+        @pipeline,
+    );
+
+    return $self->collection->aggregate(
+        \@pipeline,
+        {
+            %{ $self->options || {} },
+            cursorType => 'tailable_await',
```
Comment (xdg) [inline lib/MongoDB/ChangeStream.pm]: I think this is unnecessary unless you made other changes for it.  A `$changestream` pipeline entry automatically results in the server creating a tailable await cursor.  *Update*: I've commented later in the op where it's introduced.

File: lib/MongoDB/ChangeStream.pm [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+=head2 next
+
+    $change_stream = $collection->watch(...);
+    $change = $change_stream->next;
+
+Waits for the next change in the collection and returns it.
+
+B<Note>: This method will wait for the amount of milliseconds passed
+as C<maxAwaitTimeMS> o L<MongoDB::Collection/watch> or the default. It
```
Comment (xdg) [inline lib/MongoDB/ChangeStream.pm]:
```
s/ o / to /;
s/or the default/or the server's default wait time/
```

File: lib/MongoDB/ChangeStream.pm [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+                die $error;
+            }
+            0 # failed, cursor was rebuilt
+        };
+    }
+
+    # this differs from drivers that block indefinitely. we have to
+    # deal with the situation where no results are available.
+    if (not defined $change) {
+        return;
```
Comment (xdg) [inline lib/MongoDB/ChangeStream.pm]: This should return explicit undef, as QueryResult::next would do (ultimately from calling shift on an empty list)

File: lib/MongoDB/ChangeStream.pm [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    if (exists $change->{_id}) {
+        my $resume_token = $change->{_id};
```
Comment (xdg) [inline lib/MongoDB/ChangeStream.pm]: Let's omit the temporary.  I don't think it hurts readability.

File: lib/MongoDB/ChangeStream.pm [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+            "resume token is missing");
+    }
+}
+
+1;
+
+=head1 SYNOPSIS
+
+    $stream = $collection->watch( $pipeline, $options );
+    while (my $change = $stream->next) {
```
Comment (xdg) [inline lib/MongoDB/ChangeStream.pm]: Please wrap this in an infinite loop with a comment about how the inner loop exits when there are no changes available.

File: lib/MongoDB/Collection.pm [hunk 10]
```diff
@@ -1000,6 +1001,79 @@ sub find_one_and_update {
...
+Perform an aggregation with an implicit initial C<$changeStream> stage
+and returns a L<MongoDB::ChangeStream> result which can be used to
+iterate over the changes in the collection. This functionality is
+available since MongoDB 3.6.
+
+    my $stream = $collection->watch();
+    my $stream = $collection->watch( \@pipeline );
+    my $stream = $collection->watch( \@pipeline, \%options );
+
+    while (my $change = $stream->next) {
```
Comment (xdg) [inline lib/MongoDB/Collection.pm]: Also wrap in an infinite loop here.

File: lib/MongoDB/Op/_Aggregate.pm [hunk 11]
```diff
@@ -75,6 +88,11 @@ sub execute {
     my $options = $self->options;
     my $is_2_6 = $link->does_write_commands;
 
+    my $query_flags = {
```
Comment (xdg) [inline lib/MongoDB/Op/_Aggregate.pm]: Is this unused?

File: lib/MongoDB/Op/_Aggregate.pm [hunk 12]
```diff
@@ -61,6 +63,17 @@ has has_out => (
     isa      => Bool,
 );
 
+has maxAwaitTimeMS => (
+    is       => 'rw',
+    isa      => Num,
+);
+
+has cursorType => (
```
Comment (xdg) [inline lib/MongoDB/Op/_Aggregate.pm]: I think this is unnecessary.  For aggregation, if there's maxAwaitTimeMS, then we can conclude that it's a tailable await.

File: lib/MongoDB/Role/_CommandCursorOp.pm [hunk 13]
```diff
@@ -46,6 +46,10 @@ sub _build_result_from_cursor {
         $self->cursorType eq 'tailable_await') {
         $max_time_ms = $self->maxAwaitTimeMS if $self->maxAwaitTimeMS;
     }
+    elsif ($self->isa('MongoDB::Op::_Aggregate') &&
+        $self->cursorType eq 'tailable_await') {
```
Comment (xdg) [inline lib/MongoDB/Role/_CommandCursorOp.pm]: As I said above, I think it's sufficient to just check if isa MongoDB::Op::_Aggregate (and that maxAwaitTimeMS is true) without an extra cursor_type.

File: lib/MongoDB/Collection.pm [hunk 14]
```diff
@@ -1000,6 +1001,79 @@ sub find_one_and_update {
...
+The optional second argument is a hash reference with options:
+
+=for :list
+* C<fullDocument> - The fullDocument to pass as an option to the
+  C<$changeStream> stage. Allowed values: C<default>, C<updateLookup>.
+  Defaults to C<default>.  When set to C<updateLookup>, the change
+  notification for partial updates will include both a delta describing the
+  changes to the document, as well as a copy of the entire document that
+  was changed from some time after the change occurred.
+* C<resumeAfter> - The logical starting point for this change stream.
```
Comment (xdg) [inline lib/MongoDB/Collection.pm]: Please add to this that it must be the `_id` from a document returned from calling next on the ChangeStream object.

File: lib/MongoDB/ChangeStream.pm (hunk 7 above)
Comment (phaylon) [reply lib/MongoDB/ChangeStream.pm]: I actually had `return undef` here before, but it is flagged on `dzil test` by the auto generated `xt/author/critic.t` test.

**Comment by phaylon** [comment]
@xdg Thanks for the review! All fixes for the issues should now be in the branch.

**Comment by xdg** [comment]
LGTM. I'll squash and merge.  Thanks!

**Comment by xdg** [comment]
Squashed and merged as f7b8339e0a98551ade46d516e73d981b0d14ea1d

**Pull request closed without merging**